bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/hbasesnapshots/HBaseSnapshotInputConfiguration.java

```diff
@@ -0,0 +1,95 @@
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.mapreduce.TableInputFormat;
+import org.apache.hadoop.hbase.mapreduce.TableSnapshotInputFormat;
+import org.apache.hadoop.hbase.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.protobuf.generated.ClientProtos;
+import org.apache.hadoop.hbase.util.Base64;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.Job;
+
+/**
+ * A {@link Configuration} that could be used in {@link HadoopFormatIO} for reading HBase snapshot
+ * hosted in Google Cloud Storage(GCS) bucket via GCS connector. It uses {@link
+ * TableSnapshotInputFormat} for reading HBase snapshots.
+ */
+class HBaseSnapshotInputConfiguration {
+
+  private static final Log LOG = LogFactory.getLog(HBaseSnapshotInputConfiguration.class);
+  private static final int BATCH_SIZE = 1000;
+
+  private final Configuration hbaseConf;
+
+  /**
+   * Constructs a new top level source.
+   *
+   * @param snapshotDir The path or pattern of the file(s) to read.
+   */
+  HBaseSnapshotInputConfiguration(
+      ValueProvider<String> gcsProjectId,
+      ValueProvider<String> snapshotDir,
+      ValueProvider<String> snapshotName,
+      ValueProvider<String> restoreDir) {
+
+    Preconditions.checkArgument(
+        snapshotDir.toString().startsWith("gs://"),
+        "snapshot folder must be hosted in a GCS bucket ");
+
+    Configuration conf = HBaseConfiguration.create();
+    try {
+      conf.set("hbase.rootdir", snapshotDir.toString());
+      conf.set("fs.AbstractFileSystem.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS");
+      conf.set("fs.gs.project.id", gcsProjectId.toString());
+      conf.set("fs.defaultFS", snapshotDir.toString());
+      conf.set("google.cloud.auth.service.account.enable", "true");
+      conf.setClass(
+          "mapreduce.job.inputformat.class", TableSnapshotInputFormat.class, InputFormat.class);
+      conf.setClass("key.class", ImmutableBytesWritable.class, Writable.class);
+      conf.setClass("value.class", Result.class, Object.class);
+      ClientProtos.Scan proto = ProtobufUtil.toScan(new Scan().setBatch(BATCH_SIZE));
+      conf.set(TableInputFormat.SCAN, Base64.encodeBytes(proto.toByteArray()));
+
+      this.LOG.debug(conf);
+      Job job = Job.getInstance(conf); // creates internal clone of hbaseConf
+      TableSnapshotInputFormat.setInput(
+          job, snapshotName.toString(), new Path(restoreDir.toString()));
+      conf = job.getConfiguration(); // extract the modified clone
+    } catch (Exception e) {
+      this.LOG.fatal(e);
+    }
+    this.hbaseConf = new SerializableConfiguration(conf).get();
```

Comment:
To make this class serializable hbaseConf should be a SerializableConfiguration and the getHbaseConf should do hbaseConf.get(). 

SerializableConfiguration is a wrapper around conf, so all you are doing is wrapping the conf and unwrapping it in the same statement.